### ISSUE TITLE
Receiving duplicate events from Version 3 publisher after upgrading subscriber from Version 2 to Version 3

### ISSUE THREAD

**Issue by tmasternak** [Particular/NServiceBus.SqlServer#181]
Taskforce: @SzymonPobiega @danielmarbach @johnsimons 

Let's assume we have Endpoint A and Endpoint B both running V2 of SqlServer. Let's assume further that A is subscribing to events published by B.

In situation when Endpoint A gets updated to V3 of SqlServer transport it will start receiving duplicated messages. The root cause of this behavior is duplicate entry in messages subscription storage that will be created at Endpoint B. Before ugrade B will have ` ["EndpointA", "EventTypeX"]`  entry in subscription storage. After subscribe message from A gets processed a second entry will be stored in form `["EndpointA@schemaA", "EventTypeX"]`.

It's important to note the problem should be solved also for situation when publisher is scaled-out and instances are updated to V3 not all at once. 

The reason for this is that:
 * In V3 the address sent in subscription messages has changed from `tableName` to `tableName@schema`
 * Subscription store holds unique constraint on endpoint address (it does not interpret it in any way) and will treat `EndpointA` and `EndpointA@scheamA` as two different addresses - this btw makes a lot of sense
 * V2 endpoint already has a subscription from when Endpoint A was still running on V2

# Who's affected
All using SqlServer V2 and publish-subscribe that want to migrate to V3. 

# Impact
Because this is a blocker for V3 upgrade it is Large.

# Potential solutions

## Provide legacy subscription mode in V3

Make the `@[schema]` suffix an opt-in feature of V3. V3 would default to V2 behavior of not sending the schema name in subscribe messages. Users would be able to use the new mode (which fixes the coupling problems mentioned in #117) only if the V3 endpoint do not subscribe to any V2 endpoints.

### Problems

 * It is hard to enforce "no V2 subscriptions" rule because subscriptions can be added later (after the upgrade) and there is no way we can detect the fact that such a subscription has been added.
 * We're taking one of the new features away from majority of users who will have mixed V2 and V3 deployments until they can upgrade everything to V3.
 
## Send the schema information in a special header

Before dispatching the subscribe message, remove the schema part from the address and move it to a custom header. The V3 publisher which gets that message would be able to read that header and reconstitute the proper `endpoint@[schema]` address before passing it to the subscription infrastructure. The V2 publisher would just ignore the header.

### Problems

 * The wire format is suboptimal and we are postponing the problem for another release instead of solving it
 * The header/address manipulation requires either a pair of behaviors (which we don't want in the transport) or hacks in the receiver and dispatcher. 
 * If the publisher is scaled out and being upgraded from V2 to V3, while it is upgraded some instances might be still on V2 while the others might be on V3. If a subscribe message is picked up by a V5 endpoint, an entry **without** schema name will be inserted. If another subscribe comes (e.g. from another instance of the subscriber) and gets to a V3 instance, an entry **with** schema name will be inserted. It can be mitigated by migration process that deduplicates entries with and without the schema.

## Use native pubsub in V3

Implement a subscription store within the transport in V3 and provide `IManageSubscriptions` interface to the core. When V3 subscribes to an event it would both send a subscribe message (in V2 format, without schema name) and insert an entry (with schema name) to the transport-level subscription store. If the subscriber happens to be in V2, it would use the message. If it happens to be V3, it would use the store. 

When starting a V3 endpoint migrates subscription from a NH subscription store to the transport-level store.

### Problems

If the publisher is scaled out and being upgraded from V2 to V3 and a subscribe message comes from a V2 endpoint and is picked up by an instance which is still on V2, the entry will get added to the NH store and will be not visible for the V3 instances until the next-in-line instance is upgraded and performs the migration process. To mitigate this the migration might be triggered either periodically or by publishing.

## Teach previous versions to understand the schema name

Release a patch to NHibernate/SQL Server transport for version 4/1 and 5/2 to understand the schema name in the subscribe message. 

### Problems

 * Nobody knows how complex a task this is. One big source of complexity is usage of `Address` which is told to ignore everything after the `@` sign. 
 * This would require telling people that in order to upgrade a given endpoint X to V3 they need to upgrade all their other endpoints communicating with X to the latest patch release of NServiceBus 4/5 and SQL Transport 1/2. The issue here is that some of these endpoints might no longer be controlled by the same group who owns X.

## Deduplicate on sending in V3

Send subscribe messages in new `table@[schema]` format. A V2 publisher instance will ignore everything after `@` and only `table` will be stored in the subscription store as an address. A V3 publisher instance will store the whole address.

When V5 instance publishes a message, NHibernate subscription store deduplicates entries **after** it converts them to `Address` class so given two entries `table@[schema]` and `table` a V2 instance will publish a dispatch a single message to `table` (potentially using schema overrides via code). A V3 instance of the same publisher will get both entries from the store and, given that schema might be different than default `dbo`, it can't deduplicate in the core. SQL transport V3 would need to deduplicate transport operations in the batch based on their `QueueAddress` and message id.

### Problems

None so far.

## Migration script and triggers which deduplicate the subscription storage entries

Migration script which deduplicates all existing storage entries and adds trigger to dedup subscriptions on the fly.

### Problems

* Since migration scripts require consent with the DBAs the operational impact is much larger than depulicating directly in the dispatcher

## Hook in before SubscriptionReceiverBehavior and dedup entries by using subscription storage querying

Hook a behavior into the pipeline before SubscriptionReceiverBehavior which dedups by querying the ISubscriptionStorage
 
### Problems

* Too involved solution. Complexity is not worth it.

# Plan Of Attack [Draft]
 - [x] Update [ProductionTests](https://github.com/Particular/ProductionTests/issues/61) to newest core and implement reproduction of the error there - https://github.com/Particular/EndToEnd/pull/51
 - [x] Implement the deduplication in dispatcher - https://github.com/Particular/NServiceBus.SqlServer/pull/200
 - [ ] Verify if solution works for all the identified scenarios
 - [x] Decide what to do with the EndToEnd repro
 - [x] Documentation is not necessary since now the transport behaves like you would expect it to behave.

# Retrospective
 - Back and forth with task force members, jumping in and out because of other priorities.
 - Issue was first not prioritized and then original TF from another issue was moved over to this one.
 - Ideally we would have preferred to go with Native Pubsub but decided to go against it since we are in the stabilization phase.
- Liked the brainstorming phase and laying out the different solutions and problem they had.
- I think it is a plus that we ended up with the most pragmatic solution
- There is no well-defined way of dealing with thought-intensive tasks where there is more than one possible implementation approach. Can we call *ready* a task where there are many open options or should it be narrowed down to a single one before we can call it *ready*? The second one implies that the decision making process is carried out outside of a TF which does not sound good. On the other hand, the requirements for TF forming (multiple time zones, etc) make it super hard to schedule face-to-face meetings.

**Comment by SzymonPobiega** [comment]
Note on how to repro: run SQLNHibernateOutbox sample first for V2 and then, without clearing the DB, for V3.

**Comment by tmasternak** [comment]
Can you consider that for V6 @Particular/tf-v6-launch-plat-dev

**Comment by andreasohlund** [comment]
Based on https://github.com/Particular/PlatformDevelopment/blob/master/Guidelines/triage.md would we patch this if it where a production issue?

**Comment by tmasternak** [comment]
@andreasohlund If that was bug on production and we could patch it I would say "Yes". In my opinion it falls under `Data incorrectly updated in storage of either the user or NServiceBus` for critical patch criteria.

**Comment by andreasohlund** [comment]
I agree, do we need to spell out "duplicate messages" explicitly in that doc?

@Particular/tf-v6-launch-plat-dev @Particular/transport-maintainers based on https://github.com/Particular/NServiceBus.SqlServer/issues/181#issuecomment-199236738 I propose that we put this in scope for v6?

**Comment by danielmarbach** [comment]
Agree this should be in scope. We should not force our users to do all-in-one upgrades. This would hinder adoption of the transport and therefore backfire.

**Comment by andreasohlund** [comment]
I have added this to the scope of v6 since this is a bug that would require us to patch v6 cc @Particular/tf-v6-launch-plat-dev

**Comment by weralabaj** [comment]
There's one more problematic aspect - depending on the configuration, some endpoints can not work correctly after upgrade. E.g. using the existing SQLNHibernateOutboxEF sample, running it requires adding:
`if (queueName.Equals("Samples.SQLNHibernateOutboxEF.Receiver", StringComparison.OrdinalIgnoreCase))
                {
                    return "receiver";
                }`
inside `UseSpecificSchema` method. Otherwise it complains about non-existent endpoint, because it took the wrong schema based on the old subscription.

**Comment by mauroservienti** [comment]
what if we ship the message with the old address style, so removing everything after the @ and we ship the `schema` in the headers?
In this way we are fully compliant at the wire level with v2 of the transport, we can make v3 be able to understand both formats such as:

* endpoint@schema
* endpoint (with schema info in the headers)

And in 3 versions, say at v6 of the transport we can drop wire support with the actual v2 and gradually go back to the full address

**Comment by SzymonPobiega** [comment]
Makes sense but requires having two behaviours - one in the incoming and
one in the subscribe pipe. I wonder if we should remove the message driven
subscriptions from the core and allow transports to deal with this subject.
Otherwise each wire compat issue will require nasty hacks.

**Comment by mauroservienti** [comment]
> I wonder if we should remove the message driven subscriptions from the core and allow transports to deal with this subject.

would be very interesting, I don't think we can deal with it at this stage, maybe in the next release cycle.

Can we introduce it as a minor release? or due to the fact that it breaks some transports is considered a breaking change and thus requires a major?

**Comment by SzymonPobiega** [comment]
The SQL transport can provide `IManageSubscriptions` and then the core would switch off its subscription handling automatically. It can be done in a minor so having hacks in the pump and the dispatcher seems like a way forward for `3.0.0`.

**Comment by weralabaj** [comment]
Is SQL the only transport that has those kind of issues as far? If so then providing custom IManageSubscriptions implementation sounds reasonable.

**Comment by SzymonPobiega** [comment]
Here's the PlatDev issue that I rose to convince us to move to native pub sub in MSMQ, SQL and ASQ: https://github.com/Particular/PlatformDevelopment/issues/664.

I believe moving to native pub sub will allow us to solve this compat problem while retaining the V3/6 advantage of not having to use schema overrides on publisher side. Here's the list of case studies (which I think is exhausting all possible scenarios):

### V6 subscriber + V5 publisher

 * The subscriber sends a subscribe message and adds an entry in the transport subscription table
 * The publisher is not aware of the transport subscription table 
 * The publisher receives the message and adds an entry in its persistence subscription store
 * The publisher publishes the message
 
### V6 subscriber + V6 publisher

 * The subscriber sends a subscribe message and adds an entry in the transport subscription table
 * The publisher reads the transport subscription table
 * The publisher publishes the message
 
### V5 subscriber + V6 publisher

 * The subscriber sends a subscribe message
 * The publisher receives the message and adds an entry in its transport subscription store
 * The publisher reads the transport subscription table
 * The publisher publishes the message

### V5 publisher + non-scaled subscriber upgrading to V6

 * The subscriber, after the upgrade, sends a subscribe message and adds an entry in the transport subscription table
 * The publisher is not aware of the transport subscription table 
 * The publisher receives the message and adds an entry in its persistence subscription store
 * The publisher publishes the message
 
### V5 subscriber + non-scaled publisher upgrading to V6

 * During install time, the publisher tries to read an existing persistence subscription store and add prime the transport subscription store
 * There is no schema/catalog information in the persistence subscription store so the endpoint has to retain any schema overrides it had in V5
 * The publisher receives the message and adds an entry in its persistence subscription store
 * The publisher publishes the message
 
### V6 publisher + non-scaled subscriber upgrading to V6
 
 * The subscriber, after the upgrade, sends a subscribe message and adds an entry in the transport subscription table
 * The publisher reads the transport subscription table
 * The publisher publishes the message
 
### V6 subscriber + non-scaled publisher upgrading to V6

 * The publisher reads the transport subscription table as the subscription entry is already there (has not been used till now)
 * The publisher publishes the message
 * The publisher can get rid of schema override for all V6 subscriber as the schema/catalog information is already in the transport subscription store

### Summary

To sum up:
 * In V3 of transport (V6 of NSB) we add `IManageSubscriptions` to the transport and store subscribe entries in a SQL table that is different from V2 persistence-based subscriber table
 * V6 endpoint instances use the table while V5 continue to use the old table
 * The new table contains schema/catalog information
 * The migration process is required for upgrading the publishers in order to prime the new subscription store

The spike PR has been done. The only missing thing seems to be the migration component which seems to be as simple as `SELECT` from one table and `INSERT` to the other.

**Comment by weralabaj** [comment]
@SzymonPobiega Do we have to introduce native pub sub in all transports first to follow this path? Or we could do that only for SQL transport?

**Comment by SzymonPobiega** [comment]
@weralabaj we can do it only in SQL transport as there are no core changes required for this. The core is aware of `IManageSubscription` and uses it automatically, if provided by the transport (in place of unicast pub-sub).

**Comment by SzymonPobiega** [comment]
Also, regarding scale out, this approach I described should work because when upgrading instances one-by-one the ones still on V5 would continue to use their subscription store (and schema overrides in code) while the ones upgraded would use the new store (primed during install). Neither would contain duplicate entries.

**Comment by SzymonPobiega** [comment]
@weralabaj @tmasternak updates with information that came out during our call.

**Comment by andreasohlund** [comment]
Do we need this for beta or RC?

**Comment by tmasternak** [comment]
I would say beta - probably not. RC yes.

**Comment by SzymonPobiega** [comment]
I'd say we need it for beta. Because beta customers are the ones that are more likely to upgrade their endpoints as opposed to create new projects. They'll upgrade and start getting duplicates which does not look like a great first experience with V6.

**Comment by andreasohlund** [comment]
@Particular/training-maintainers please make a call on https://github.com/Particular/NServiceBus.SqlServer/issues/181#issuecomment-208268540

**Comment by DavidBoike** [comment]
@andreasohlund **Training** maintainers?

**Comment by andreasohlund** [comment]
fixed, sorry :)

**Comment by tmasternak** [comment]
> They'll upgrade and start getting duplicates which does not look like a great first experience with V6

That's a valid argument. I agree we should have this for beta.

**Comment by andreasohlund** [comment]
Ok, @Particular/platform-dev-squad I'm moving this one to next in line. Let's transfer the task force from https://github.com/Particular/PlatformDevelopment/issues/732 and start this one instead?

**Comment by SzymonPobiega** [comment]
:+1:

**Comment by andreasohlund** [comment]
After talking to @SzymonPobiega we came to the conclusion that since we don't have a clear idea on how to solve this it shouldn't be in progress and should not have a task force. @SzymonPobiega will work (and pull in relevant people as needed) in order to come up with a POA. Once we have that we can staff this properly and complete it.

Moving this back to next in line and removing @danielmarbach and @petersgiles from the taskforce

**Comment by weralabaj** [comment]
Since Szymon is working on it then why isn't that in progress?
I understand that it's problematic that we have a 1-person TF, but I don't think this is the right solution because we lose visibility into what's going on.

**Comment by adamralph** [comment]
We discussed offline and agreed that @SzymonPobiega will continue to explore the issue and get a TF together before it's ready to go into progress again (WIP limit allowing).

**Comment by danielmarbach** [comment]
Why can't the task force do the thinking to get this ready? So a dedicated
TF to get the PoA for this conplex issue. One single person (sorry szymon)
might not have the necessary brainwave to PoA convertion power available

**Comment by SzymonPobiega** [comment]
@danielmarbach if I understand this correctly, this is more a matter of taste. I don't blame the squad, they are doing what they are expected to. It is clear that I won't be able to make a decision myself and there will be *a group* of people involved. Because calling that group a *task force* carries a lot of assumptions, constraints etc. I am fine with calling it a group. You are more than welcome to be part of that group.

**Comment by SzymonPobiega** [comment]
@danielmarbach @petersgiles would you like to (re)join the TF for here? I've did some additional research and I am confident this is now in the ready state.

**Comment by SzymonPobiega** [comment]
@tmasternak @weralabaj @mauroservienti can you, as :chicken: go through the possible solutions described in this issue and add your comments here? If we can do this async, it would be easier to manage. Worst case I can set up a call.

**Comment by SzymonPobiega** [comment]
@danielmarbach Should we add #198 to the plan of attack here?

**Comment by danielmarbach** [comment]
@tmasternak @SzymonPobiega @weralabaj is the motivation to use the schema for v3 written down somewhere?

**Comment by danielmarbach** [comment]
@SzymonPobiega yes I'll do

**Comment by mauroservienti** [comment]
After reading all @SzymonPobiega proposals my preference is for `native pub/sub in V3`. The only cons is at upgrade time in a scenario where

1. multiple schemes are in place
* the endpoint is scaled out
* the environment must be upgraded with no downtime using side-by-side deployments
* a subscription comes in during the upgrade and is handled by a V2 instance

if I recall correctly we said, in Bangkok, that 3 is not really a supported scenario, if this is the case then `native pub/sub in V3` is the optimal solution, IMO, the real con is that we must introduce it in V3.0 and cannot be done in a minor.

**Comment by tmasternak** [comment]
@danielmarbach #116 #117. The main reason to put it in address is to avoid hidden coupling.

**Comment by mauroservienti** [comment]
And to make ServiceControl able to work with the SQL transport when using the multi scheme feature

**Comment by danielmarbach** [comment]
@mauroservienti although native pub sub would be our preference as well we discussed that it is too involved right now. Even when we introduce native pub sub (which might have other unforeseen consequences) we could potentially end up with v2 and v3 nodes that need to work together. So in that case how would you migrated on the fly from the NH storage to the transport storage without risking to walk into the same problems of duplicated messages. That's why we said Nay to native pub sub for now.

**Comment by mauroservienti** [comment]
@danielmarbach as I said it this:

> the environment must be upgraded with no downtime using side-by-side deployments

is a supported scenario, and AFAIK it is not, than I agree with you.

**Comment by danielmarbach** [comment]
the problem is all solutions we discussed which involved migrations were actually an all or nothing upgrade scenario. I don't think we can agree to this, do you?

**Comment by SzymonPobiega** [comment]
I think we can have native pub sub without the need to upgrade all. Need to
go through all scenarios.

**Comment by danielmarbach** [comment]
@SzymonPobiega let's have a discussion.

**Comment by tmasternak** [comment]
As a :chicken: I going to say that for me the most promising is the `Deduplication on sending in V3` and `Adding triggers on the database level to subscription storage` (btw. the last one should probably be added into description).

**Comment by danielmarbach** [comment]
Today was yak shaving day. Had to clean up existing prod tests, move them over to the end2end test repo and then based on that write the reproduction test. The reproduction test is now up and running as a PR on the end2end test repo and linked against this issue.

**Comment by SzymonPobiega** [comment]
* Updated the PoA
 * Modified the test to properly wait
 * Implemented a WIP version of dispatcher deduplication (that does not rule out other options yet at this stage )

**Comment by danielmarbach** [comment]
We decided to leave the test in the end2end test repo. We know that the value of this test will diminish over time. But it is currently a good example of how to write end 2 end test with upgrading endpoints.

**Comment by danielmarbach** [comment]
Added retro section and notes about documentation

**Comment by SzymonPobiega** [comment]
Scenario-based verification of implemented solution (dispatcher deduplication):

* [ ] V5 publisher + new V6 subscriber
* [ ] V5 subscriber + new V6 publisher
* [ ] new V6 subscriber + new V6 publisher
* [ ] V5 publisher upgrading to V6 + V5 subscriber upgrading to V6
* [ ] V5 publisher  + non-scaled subscriber upgrading to V6
* [ ] V5 subscriber + non-scaled publisher upgrading to V6
* [ ] V6 publisher  + non-scaled subscriber upgrading to V6
* [ ] V6 subscriber + non-scaled publisher upgrading to V6
* [ ] V5 subscriber + scaled out publisher upgrading to V6 (one-by-one)

**Comment by andreasohlund** [comment]
Could this have negative impact on performance? (any specific tests that we need to do)

**Comment by johnsimons** [comment]
I can help with the testing tomorrow

On Mon, 18 Apr 2016 at 18:29 Szymon Pobiega <notifications@github.com>
wrote:

> Scenario-based verification of implemented solution (dispatcher
> deduplication):
>
>    - V5 publisher + new V6 subscriber
>    - V5 subscriber + new V6 publisher
>    - new V6 subscriber + new V6 publisher
>    - V5 publisher upgrading to V6 + V5 subscriber upgrading to V6
>    - V5 publisher + non-scaled subscriber upgrading to V6
>    - V5 subscriber + non-scaled publisher upgrading to V6
>    - V6 publisher + non-scaled subscriber upgrading to V6
>    - V6 subscriber + non-scaled publisher upgrading to V6
>    - V5 subscriber + scaled out publisher upgrading to V6 (one-by-one)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus.SqlServer/issues/181#issuecomment-211266563>
>
-- 

*John Simons*
*Engineer, Particular Software*

Email*:*  john.simons@particular.net
Web:    www.particular.net

**Comment by danielmarbach** [comment]
@andreasohlund technically not much. The collections are dedupped in memory and usually pretty small.

**Comment by johnsimons** [comment]
So it seems all the testing has been completed by @SzymonPobiega. Anything else or can this one be closed ?

**Comment by SzymonPobiega** [comment]
@johnsimons yeah. The only thing that's left is my retrospective node. I'll do it now and I am closing this one.

**Comment by SzymonPobiega** [comment]
Now that #205 is closed, I can properly close this one. I'll add retro entry mentioning the failure.
